doc/developer/design/20230411_envelope_upsert_order_by_timestamp.md

```diff
@@ -0,0 +1,110 @@
+- Feature name: Envelope upsert with order by timestamp (limited scope)
+- Associated: https://github.com/MaterializeInc/materialize/issues/16512
+
+# Summary
+
+- Support user specified order by in Kafka upsert sources in a limited capacity. If users want to specify custom ordering which is semantically meaningful to them, there's currently no easy way to do that. They could potentially use `ENVELOPE NONE` along with a `DISTINCT ON` view to get similar results, but it's cumbersome and lacks the automatic compaction on keys with normal `ENVELOPE UPSERT`.
+- Currently, will limit the scope to allow only `ORDER BY` the `TIMESTAMP` and the `OFFSET` metadata fields. Depending upon user feedback this can later be expanded to support expressions and extractions from other columns like `HEADERS`.
+
+# Scope
+
+The scope will be limited to only allow `ORDER BY ( TIMESTAMP, OFFSET )` (ascending order).
+
+- In a key-compacted Kafka topic, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to borrow that tombstone semantic for our more flexible upsert source, it would make sense to limit the `ORDER BY` columns to only the metadata columns.
+- Limiting it to only `TIMESTAMP` and `OFFSET` because
+    - `OFFSET` is already default. We are making it explicit here and it's going to be required that the user always mention `OFFSET` as one of the columns in ORDER BY.
+    - `PARTITION`, `KEY`, `TOPIC`, would not make for meaningful orderings:
+        - The `PARTITION` of a record is typically derived from its key, and so is the same for every record with a given key.
+        - `KEY` is, by definition, the same for every record with a given key.
+        - `TOPIC` is the same for every record in the topic.
+    - `HEADERS` by itself would not also make a good order by clause (it’s a json blob), but later we can probably allow expressions so that values could be extracted from it.
+- Limiting to only ascending order because no users have expressed a need for descending.
+
+# Explanation
+
+## Proposal
+
+Make it possible with unsafe-mode to provide custom ORDER BY clause which will override the default ordering by offset. The given ORDER BY identifiers should refer to only the included TIMESTAMP and OFFSET metadata columns. It will be required that the user always mention OFFSET as one of the columns for the ORDER BY clause.
+
+The option will be part of the `ENVELOPE UPSERT` clause with the following grammar:
+
+`ENVELOPE UPSERT [(ORDER BY (<expr>) [ASC])]`
```

Comment:
With the change to support multiple columns, this should be updated to:

```
ENVELOPE UPSERT [(ORDER BY (<expr> [ASC], ...))]
```

To match the way this works for `SELECT`, the `ASC` should be optionally attached to _each_ expression.

Reply:
Ah I see what you mean. I saw `ORDER BY (col1, col2) ASC` is legal as well in select statements, so had kept that (I guess it's the same thing as you mentioned where the entire tuple is the expression). I will change it to ordering per expression.